### DISCUSSION TITLE
Different wallpaper per group

### DISCUSSION THREAD

**Discussion by raiguard** [qtile/qtile#2782] [Q&A] [answered]
I would really like to have a different wallpaper per group in QTile. I had that set in PaperWM when I still used Gnome, and it makes it very easy to remember which workspace you're in, and allows me to see a ton of different wallpapers throughout the day if I so like.

**Comment by m-col** [comment]
Was that a PaperWM feature or a Gnome feature?

**Reply by raiguard** [reply]
It was a PaperWM feature.

**Comment by m-col** [comment]
I think your best bet is to use the `setgroup` hook and set the wallpaper when it is fired:
```py
from libqtile import qtile, hook

groups = [...]  # Let's say you have 10 groups
wallpapers = [...]  # Then list 10 wallpapers

@hook.subscribe.setgroup
def set_wallpaper():
    wallpaper = wallpapers[qtile.groups.index(qtile.current_group)]
    qtile.paint_screen(qtile.current_screen, wallpaper, mode="fill")
```

Untested, but I think that would work with a bit of additional configuration.

**Reply by raiguard** [reply]
Thank you for the example!

It took me a long time to get around to making it work, because I do not know python well whatsoever, and it wasn't working for some reason. Turns out I just had the paths incorrect, and had to use `os.path.expanduser` on them. Now it works!

However, there are a few things:
- It only changes it for the current monitor - the other monitor's workspace changes but its background stays the same
- There is a significant delay and noticeable lag spike when switching wallpapers in some cases

For the first one, I assume I would just have to iterate over all screens during the hook and update all wallpapers every time. But that will probably exacerbate the second issue. I've noticed that it lags especially hard if I have [kitty](https://github.com/kovidgoyal/kitty) open on that workspace.

Also, `fill` doesn't appear to center the wallpaper. Is there a way to make it do that?

https://user-images.githubusercontent.com/3515394/138392481-f9cbadcc-d1bc-4e10-83bc-a7feb1979d62.mp4

**Comment by m-col** [comment, marked as answer]
> It only changes it for the current monitor - the other monitor's workspace changes but its background stays the same

This might be a bug - I remember recently there was a discussion about problematic wallpaper handling when using multiple monitors. How are your monitors set up - left and right, or top and bottom? Possibly related: https://github.com/qtile/qtile/pull/2839

> There is a significant delay and noticeable lag spike when switching wallpapers in some cases

Hmm not ideal. The main thing that comes to mind is that the code that paints the wallpaper does no caching at all because usually it's a single call at startup. This means that each time your wallpaper is being painted the image data is being loaded again from file, and that's a likely culprit in the delay. It might not be too hard to implement this in your config though, so maybe give that a go. To do this, I'd suggest:

1. Copy the code from `libqtile.backend.x11.xcbq.Painter` into your config.
2. You want to keep the code in `__init__()` but you may need to update some of the variables so that they can find other things inside `xcbq`.
3. Either in `__init__` or a new method, pass it all of your wallpapers and use `cairocffi.pixbuf.decode_to_image_surface` to load all of them into memory and store these on the instance. Do this in your config as it is loaded, and then all of those images can be taken directly from memory when used.
4. Then just take the rest of `Painter.paint()` but adjust the top of it so that it's using the correct image data.
5. In your `setgroup` hook, directly use its `paint` method instead of going through `Qtile.paint_screen`.

This also might get a slight speedup by not doing `__init__` each time because it has to make a connection to the X server, and this way you just make one and it persists throughout the session.

Anyway I hope this helps, let me know if you want any help!

**Reply by raiguard** [reply]
My monitors are side by side, both 1440p. Sometimes, I also enable output to my 4k TV, like so:

![image](https://user-images.githubusercontent.com/3515394/138559425-b911ca24-5026-428c-9260-d7615d83e969.png)

However, when I do so, the wallpapers get all jacked up:

(edit: just noticed that the referenced #2839 is directly related to this!)

![20211023_075848](https://user-images.githubusercontent.com/3515394/138559598-9ac82f01-6af5-4b1d-b263-ef59c7b130e4.jpg)

The wallpaper workspace change script simply does not work on this screen. Not to mention that it switches my primary screen to the TV, but I can fix that with a bit of scripting.

I will try the caching code and let you know what I find.

My configuration, including the wallpapers that I use, can be found [here](https://github.com/raiguard/dotfiles).

**Reply by raiguard** [reply]
It's been a little while. Since I posted this, I switched to KDE Plasma, then came back to Qtile. I'm on a brand-new Arch install with Qtile as the only WM / DE, and it's really nice.

I'm no longer using a different wallpaper per group, so I marked your response as the answer, since it did work (despite the performance issues). The wallpaper issue persists, but I will continue discussion on that in #2839, after I test to see if that PR fixes my issue.

Thanks!
